src/main/java/com/google/devtools/build/lib/exec/SymlinkTreeStrategy.java

```diff
@@ -131,11 +132,14 @@ private static void createOutput(
       SymlinkTreeAction action, ActionExecutionContext actionExecutionContext, Path inputManifest)
       throws EnvironmentalExecException {
     Path outputManifest = actionExecutionContext.getInputPath(action.getOutputManifest());
-    // Link output manifest on success. We avoid a file copy as these manifests may be
-    // large. Note that this step has to come last because the OutputService may delete any
-    // pre-existing symlink tree before creating a new one.
+    // Copy output manifest on success. We can't use a symlink or hardlink here because
+    // SymlinkTreeAction is executed for its side effect (the symlink tree creation) and could
+    // erroneously be considered up-to-date if the input manifest changed back to a different state
+    // that had previously been recorded in the action cache.
+    // Note that this step has to come last because the OutputService may delete any pre-existing
+    // symlink tree before creating a new one.
     try {
-      outputManifest.createSymbolicLink(inputManifest);
+      FileSystemUtils.copyFile(inputManifest, outputManifest);
```

Comment:
What about the corresponding logic in `RunfilesTreeUpdater#updateRunfilesTree`? Should we also make a copy there?

Reply:
The fast path is actually never taken when the output manifest is a symlink, so it may be more efficient to make it a copy. My new tests don't fail if I keep it as a symlink, but I'm not sure what happens if you flip `--build_runfile_links` somewhere in the chain of builds. That may start to fail if we drop the `NotifyOnActionCacheHit` logic some day. I will make it a copy just to be safe and consistent.